Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,280 @@
+import Mathlib.Data.Set.Functor
+import Mathlib.Lean.Expr.ExtraRecognizers
+
+/-!
+# Sets in subtypes
+
+This file is about sets in `Set A` when `A` is a set.
+
+It defines notation `↓∩` for sets in a type pulled down to sets in a subtype, as an inverse
+operation to the coercion that lifts sets in a subtype up to sets in the ambient type.
+
+This module also provides lemmas for `↓∩` and this coercion.
+
+## Notation
+
+Let `α` be a `Type`, `A B : Set α` two sets in `α`, and `C : Set A` a set in the subtype `↑A`.
+
+- `A ↓∩ B` denotes `(Subtype.val ⁻¹' B : Set A)` (that is, `{x : ↑A | ↑x ∈ B}`).
+- `↑C` denotes `Subtype.val '' C` (that is, `{x : α | ∃ y ∈ C, ↑y = x}`).
+
+This notation is scoped to the `Set.Notation` namespace.
+To enable it, use `open Set.Notation`.
+
+
+## Naming conventions
+
+Theorem names refer to `↓∩` as `preimage_val`.
+
+## Tags
+
+subsets
+-/
+
+open Set
+
+universe u
+
+variable {α β : Type u} {A B C : Set α} {D E : Set A}
+variable {S : Set (Set α)} {T : Set (Set A)} {i : β → Set α} {j : β → Set A}
+
+namespace Set.Notation
+
+/--
+Given two sets `A` and `B`, `A ↓∩ B` denotes the intersection of `A` and `B` as a set in `Set A`.
+
+The notation is short for `((↑) ⁻¹' B : Set A)`, while giving hints to the elaborator
+that both `A` and `B` are terms of `Set α` for the same `α`.
+This set is the same as `{x : ↑A | ↑x ∈ B}`.
+-/
+scoped notation3 A:67 " ↓∩ " B:67 => (Subtype.val ⁻¹' (B : type_of% A) : Set (A : Set _))
+
+open Lean PrettyPrinter Delaborator SubExpr in
+/--
+Sets of a subtype coerced to the ambient type are denoted with `↑`.
+-/
+@[scoped delab app.Set.image]
+def delab_set_image_subtype : Delab := do
+  let #[α, _, f, _] := (← getExpr).getAppArgs | failure
+  guard <| f.isAppOfArity ``Subtype.val 2
+  let some _ := α.coeTypeSet? | failure
+  let e ← withAppArg delab
+  `(↑$e)
+
+end Set.Notation
+
+namespace Set
+
+open Set.Notation
+
+lemma preimage_val_eq_univ_of_subset (h : A ⊆ B) : A ↓∩ B = univ := by
+  ext x
+  simp only [mem_univ, iff_true]
+  exact h x.2
+
+lemma preimage_val_subset_preimage_val_iff : A ↓∩ B ⊆ A ↓∩ C ↔ A ∩ B ⊆ A ∩ C := by
+  constructor
+  · rintro h x ⟨hxA, hxB⟩
+    constructor
+    · exact hxA
+    · specialize h ?_
+      · exact ⟨x, hxA⟩
+      · exact hxB
+      exact h
+  · rintro h ⟨x, hxA⟩ hx
+    specialize h ?_
+    · exact x
+    · exact ⟨hxA, hx⟩
+    exact h.2
+
+lemma preimage_val_eq_iff : A ↓∩ B = A ↓∩ C ↔ A ∩ B = A ∩ C := by
+  simp only [subset_antisymm_iff, preimage_val_subset_preimage_val_iff, subset_inter_iff,
+    inter_subset_left, true_and]
+
+lemma preimage_val_sUnion : A ↓∩ (⋃₀ S) = ⋃₀ { (A ↓∩ B) | B ∈ S } := by
+  ext x
+  simp only [preimage_sUnion, mem_iUnion, mem_preimage, exists_prop, mem_sUnion, mem_setOf_eq,
+    exists_exists_and_eq_and]
+
+@[simp]
+lemma preimage_val_iInter : A ↓∩ (⋂ (B : β), i B) = ⋂ (B : β), A ↓∩ i B := by
+  exact preimage_iInter
+
+lemma preimage_val_sInter : A ↓∩ (⋂₀ S) = ⋂₀ { (A ↓∩ B) | B ∈ S } := by
```

Comment:
I think this should be proven for the case of a general preimage, since the proof doesn't mention `val`.

Reply:
(or maybe we should just have a lemma that says `⋂₀ { f B | B ∈ S } = ⋂ B ∈ S, f B`, since the latter is the preferred spelling)

Reply:
You mean for arbitrary `f : Set α → Set β` ?

Reply:
I think this lemma is fine to have, at least to specialize the notation and for discoverability.

@eric-wieser `Set.preimage_sInter` is that general lemma.

I guess the question is whether the RHS should instead be `⋂ B ∈ S, (A ↓∩ B)`. That way the proof of this lemma would be exactly `Set.preimage_sInter`.

Reply:
> I guess the question is whether the RHS should instead be `⋂ B ∈ S, (A ↓∩ B)`. That way the proof of this lemma would be exactly `Set.preimage_sInter`.

I think the RHS should indeed be changed (here and elsewhere)

Reply:
So the general idea is to avoid `⋂₀` in general, and change it to `⋂` (and same for unions)? 

I have the opinion that it is nice to have lemmas that allow one to rewrite expressions with `⋂₀`  into other expressions with `⋂₀`, without converting them into `⋂`. It reflects better some of the arguments we use on paper.

As @kmill commented, there is already a lemma that gives the other equality.

Reply:
> I have the opinion that it is nice to have lemmas that allow one to rewrite expressions with `⋂₀` into other expressions with `⋂₀`, without converting them into `⋂`. It reflects better some of the arguments we use on paper.

If you feel strongly about this, then the resolution would be to change `Set.preimage_sInter` to have this statement instead of the current one (in a separate PR, and after discussion on zulip). We should try to avoid making difference choices to the same question in different files.

If we really were to change it, I'd argue that the RHS of the lemma should be `⋂₀ ((A ↓∩ .) '' S)` since that's the canonical spelling of a set image.

Reply:
what do you think @kmill ?

Reply:
Would it make sense to have both? Call this variant `preimage_val_sInter_eq_sInter`?

Reply:
I just added it.